neuralmonkey/dataset.py

```diff
@@ -0,0 +1,556 @@
+                "Preprocessed series '{}', source series '{}'")
+        iterators[s_name] = _make_sl_iterator(source, preprocessor)
+
+    # Finally, dataset-level preprocessors.
+    for s_name, func in prep_dl.items():
+        iterators[s_name] = _make_dl_iterator(func)
+
+    output_dict = None
+    if outputs is not None:
+        output_dict = {s_name: (path, writer)
+                       for s_name, path, writer
+                       in [_normalize_outputspec(out) for out in outputs]}
+
+    return Dataset(name, iterators, output_dict, lazy, buffer_size, shuffled)
+# pylint: enable=too-many-locals,too-many-branches
+
+
+class Dataset:
+    """Buffered and batched dataset.
+
+    This class serves as collection of data series for particular encoders and
+    decoders in the model.
+
+    Dataset has a number of data series, which are sequences of data (of any
+    type) that should have the same length. The sequences are loaded in a
+    buffer and can be loaded lazily.
+
+    Using the `batches` method, dataset yields batches, through which the data
+    are accessed by the model.
+    """
+
+    def __init__(self,
+                 name: str,
+                 iterators: Dict[str, Callable[[], Iterator]],
+                 outputs: Dict[str, Tuple[str, Writer]] = None,
+                 lazy: bool = False,
+                 buffer_size: int = None,
+                 shuffled: bool = False) -> None:
+        """Construct a new instance of the dataset class.
+
+        Do not call this method from the configuration directly. Instead, use
+        the `from_files` function of this module.
+
+        The dataset iterators are provided through factory functions, which
+        return the opened iterators when called with no arguments.
+
+        Arguments:
+            name: The name for the dataset.
+            iterators: A series-iterator generator mapping.
+            lazy: If False, load the data from iterators to a list and store
+                the list in memory.
+            buffer_size: If lazy, use this buffer size for pre-loading data.
+            shuffled: Whether to shuffle the buffer during batching.
+        """
+        self.name = name
+        self.iterators = iterators
+        self.outputs = outputs
+        self.lazy = lazy
+        self.buffer_size = buffer_size
+        self.shuffled = shuffled
+        self.length = None
+
+        if not self.lazy:
+            data = {s_name: list(it())
+                    for s_name, it in self.iterators.items()}
+
+            # Check whether all loaded series have the same length
+            length_dict = {
+                s_name: len(s_data) for s_name, s_data in data.items()}
+            if len(set(length_dict.values())) > 1:
+                raise ValueError("Lengths of data series do not match: {}"
+                                 .format(str(length_dict)))
+
+            self.length = next(iter(length_dict.values()))
```

Comment:
Is the next(iter()) syntax necessary? Shouldn't length_dict.values()[0] be the same?

Reply:
```ipython
In [1]: a = {"barack": 67, "hussein": 12, "obama": 32}

In [2]: a.values()[0]
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-53f4ff1337fa> in <module>()
----> 1 a.values()[0]

TypeError: 'dict_values' object does not support indexing
```

Reply:
https://stackoverflow.com/questions/3097866/access-an-arbitrary-element-in-a-dictionary-in-python

Reply:
Oh, now I see it.